### DISCUSSION TITLE
[Bar Chart] Set max width of a bar

### DISCUSSION THREAD

**Discussion by williaster** [airbnb/visx#1035] [Q&A] [answered]
The following was originally posted as an issue by @wisyr in #1032, moving it to a discussion.

----

tl;dr - Is there a way to set a max width of a bar in charts but still keep it dynamic? 

The issue is related with bar width. If there's enough of data, a single bar looks just okey.
![image](https://user-images.githubusercontent.com/20928453/105636114-f846d980-5e66-11eb-80a7-9254941d70b2.png)

But if there's no enough data, the bars are getting too wide.
![image](https://user-images.githubusercontent.com/20928453/105636160-4065fc00-5e67-11eb-9e64-55ca3268a6fe.png)

Is there a way to somehow set a max width of a single bar? But still keep it dynamic so it is responsive when chart is resizing.

Playground: https://codesandbox.io/s/holy-pine-i5sp2?file=/Example.tsx

Thank you for such an outstanding library!

**Comment by williaster** [comment, marked as answer]
@wisyr `barWidth` for a `band` type scale is determined by the available chart `width` (set as the scale's `range`) and the `padding` (`paddingInner` + `paddingOuter`) set on the `scale`. This is described in detail and depicted in [`d3-scale`'s docs](https://github.com/d3/d3-scale#band-scales):

<img src="https://user-images.githubusercontent.com/4496521/105773251-77481a80-5f18-11eb-9727-25749763da78.png" width="700" />

So to get this behavior, you need to compute the necessary padding (using the available chart `width` and number of bars) in order to achieve your desired maximum width. Here's a [sandbox that demonstrates the gist of it](https://codesandbox.io/s/wonderful-herschel-61i0b?file=/Example.tsx).

<img src="https://user-images.githubusercontent.com/4496521/105773378-a5c5f580-5f18-11eb-83ec-7981dfd72217.png" width="700" />

Note you can get different behavior by setting `scale.padding` (above, in sandbox) vs `scale.paddingInner` (below):

<img src="https://user-images.githubusercontent.com/4496521/105773643-10773100-5f19-11eb-91f7-987b9535cf38.png" width="700" />

Hope that helps, let me know if I can clarify anything further.

**Reply by valtism** [reply]
This is great, but I think it is slightly off in the calculations (even though they seem to come out correct). When I was implementing this with bar groups, I had to take into account the number of bars per group and rejig the calculation.

```js
const barsPerGroup = dataGroups[0].length
const numBars = dateRange.length
const barWidth = barsPerGroup * maxBarWidth
const barsWidthTotal = numBars * barWidth
const padding = (chartWidth - barsWidthTotal) / (numBars + 1)
const band = padding + barWidth
const paddingRatio = padding / band
```

Even then I would get bars which were slightly off in terms of size.

**Reply by rajatdangat07** [reply]
Here is a solution for grouped bar chart. Here's a [sandbox demo](https://codesandbox.io/s/wizardly-ellis-eun1w?file=/Example.tsx).

**Comment by wisyr** [comment]
Dear @williaster 
Absolutely incredible. Thank you for your time. Works like a charm.
Cheers!
